pkg/shortnames/shortnames_test.go

```diff
@@ -0,0 +1,473 @@
+package shortnames
+
+import (
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/pkg/sysregistriesv2"
+	"github.com/containers/image/v5/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestIsShortName(t *testing.T) {
+	tests := []struct {
+		input       string
+		isShortName bool
+		mustFail    bool
+	}{
+		// SHORT NAMES
+		{"fedora", true, false},
+		{"fedora:latest", true, false},
+		{"library/fedora", true, false},
+		{"library/fedora:latest", true, false},
+		{"busybox@sha256:d366a4665ab44f0648d7a00ae3fae139d55e32f9712c67accd604bb55df9d05a", true, false},
+		// !SHORT NAMES
+		{"quay.io/fedora", false, false},
+		{"docker.io/fedora", false, false},
+		{"docker.io/library/fedora:latest", false, false},
+		{"localhost/fedora", false, false},
+		{"localhost:5000/fedora:latest", false, false},
+		{"example.foo.this.may.be.garbage.but.maybe.not:1234/fedora:latest", false, false},
+		{"docker.io/library/busybox@sha256:d366a4665ab44f0648d7a00ae3fae139d55e32f9712c67accd604bb55df9d05a", false, false},
+		{"docker.io/fedora@sha256:d366a4665ab44f0648d7a00ae3fae139d55e32f9712c67accd604bb55df9d05a", false, false},
+		// INVALID NAMES
+		{"", false, true},
+		{"$$$", false, true},
+		{"::", false, true},
+		{"docker://quay.io/library/foo:bar", false, true},
+		{" ", false, true},
+	}
+
+	for _, test := range tests {
+		res, _, err := isShortName(test.input)
+		if test.mustFail {
+			require.Error(t, err, "%q should not be parseable")
+			continue
+		}
+		require.NoError(t, err, "%q should be parseable")
+		assert.Equal(t, test.isShortName, res, "%q", test.input)
+	}
+}
+
+func TestNormalizeUserInput(t *testing.T) {
+	tests := []struct {
+		input      string
+		normalized string
+		isTagged   bool
+		isDigested bool
+	}{
+		// Neither tags nor digests
+		{"fedora", "fedora", false, false},
+		{"repo/fedora", "repo/fedora", false, false},
+		{"registry.com/fedora", "registry.com/fedora", false, false},
+		// Tags
+		{"fedora:tag", "fedora", true, false},
+		{"repo/fedora:tag", "repo/fedora", true, false},
+		{"registry.com/fedora:latest", "registry.com/fedora", true, false},
+		// Digests
+		{"fedora@sha256:d366a4665ab44f0648d7a00ae3fae139d55e32f9712c67accd604bb55df9d05a", "fedora", false, true},
+		{"repo/fedora@sha256:d366a4665ab44f0648d7a00ae3fae139d55e32f9712c67accd604bb55df9d05a", "repo/fedora", false, true},
+		{"registry.com/fedora@sha256:d366a4665ab44f0648d7a00ae3fae139d55e32f9712c67accd604bb55df9d05a", "registry.com/fedora", false, true},
```

Comment:
It would be nice to add a “repo:tag@digest” test case; AFAICS the current code works but I have already started considering a refactoring before noticing that it would break that case.